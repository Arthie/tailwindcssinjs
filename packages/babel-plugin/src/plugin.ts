// To do: add package version to the cache //hash of package.json???
// Check if config is changed + save hash of tailwind.config.js
// @future https://github.com/babel/babel/pull/11741 + maybe depend on generated css???
// check if output css file exists and warn if not
// merge screen atrules?
import Babel, { PluginObj, PluginPass } from "@babel/core";
import initTwClassesUtils, {
  ParsedTwClass,
} from "@tailwindcssinjs/class-composer";
import fs from "fs";
import crypto from "crypto";
import {
  createTwClassDictionary,
  getGenerateTwClassSubstituteRoot,
  requireTailwindConfig,
  resolveTailwindConfigPath,
  tailwindData,
  TwClassDictionary,
} from "@tailwindcssinjs/tailwindcss-data";
import corePlugins from "tailwindcss/lib/corePlugins";

function getHash(source: string) {
  return crypto.createHash("md5").update(source).digest("hex");
}

function getTailwindConfig(config: string) {
  try {
    const resolvedConfigPath = resolveTailwindConfigPath(config);
    return requireTailwindConfig(resolvedConfigPath);
  } catch (err) {
    return requireTailwindConfig(); //returns default config
  }
}

interface TailwincssinjsCacheEntry {
  sourcePath: string;
  sourceHash: string;
  twClasses: string[];
}

function initCache(path: string) {
  const cachePath = path + "/.tailwindcssinjs/cache.json";

  const getCache = (): TailwincssinjsCacheEntry[] => {
    if (fs.existsSync(cachePath)) {
      const file = fs.readFileSync(cachePath, { encoding: "utf8" });
      return JSON.parse(file);
    } else {
      if (!fs.existsSync(path + "/.tailwindcssinjs")) {
        fs.mkdirSync(path + "/.tailwindcssinjs");
      }
      fs.writeFileSync(cachePath, "[]");
      return [];
    }
  };

  const validateCache = (cache: TailwincssinjsCacheEntry[]) => {
    const validatedCache: TailwincssinjsCacheEntry[] = [];
    for (const entry of cache) {
      if (fs.existsSync(entry.sourcePath)) {
        const file = fs.readFileSync(entry.sourcePath, { encoding: "utf8" });
        const fileHash = getHash(file);
        if (entry.sourceHash === fileHash) {
          validatedCache.push(entry);
        } else {
          console.log(
            "hash not right",
            entry.sourcePath,
            fileHash,
            entry.sourceHash
          );
        }
      } else {
        console.log("file is gone", entry.sourcePath);
      }
    }
    return validatedCache;
  };

  const updateCache = (
    cache: TailwincssinjsCacheEntry[],
    updateEntry: TailwincssinjsCacheEntry
  ) => {
    const index = cache.findIndex(
      (entry) => entry.sourcePath === updateEntry.sourcePath
    );
    if (index === -1) {
      cache.push(updateEntry);
    } else {
      cache[index] = updateEntry;
    }
    return cache;
  };

  const writeCache = (cache: TailwincssinjsCacheEntry[]) => {
    fs.writeFileSync(cachePath, JSON.stringify(cache));
  };

  return {
    getCache,
    validateCache,
    updateCache,
    writeCache,
  };
}

let $configHash = "";
let $generateTwClassSubstituteRoot:
  | ReturnType<typeof getGenerateTwClassSubstituteRoot>
  | undefined;
let $twClassesUtils: ReturnType<typeof initTwClassesUtils> | undefined;
let $twClassDictionary: TwClassDictionary | undefined;
let $compare: any | undefined;
let $generatedTwClassesCSS: { [key: string]: string } = {};

let baseCSS = "";
const commentCSS = `/* Generated by @tailwindcssinjs/babel-plugin */`;

export default function (
  babel: typeof Babel,
  config: {
    cache?: boolean;
    includeBase?: boolean;
    config?: string;
    output?: string;
  },
  workingPath: string
): PluginObj<PluginPass> {
  const { types: t } = babel;

  const enableCache = config.cache ?? true;
  let cacheUtils: ReturnType<typeof initCache> | undefined;
  let cache: TailwincssinjsCacheEntry[] | undefined;
  if (enableCache) {
    cacheUtils = initCache(workingPath);
    cache = cacheUtils.validateCache(cacheUtils.getCache());
  }

  const outputPath = config.output;
  const configPath = config.config ?? "./tailwind.config.js";
  const includeBase = config.includeBase ?? true;

  return {
    name: "tailwindcssinjs",
    visitor: {
      ImportDefaultSpecifier(path, state) {
        if (path.parent.type !== "ImportDeclaration") return;
        if (path.parent.source.value !== "@tailwindcssinjs/css") return;

        const importDefaultName = path.node.local.name;
        const referencePaths =
          path.scope.bindings[importDefaultName].referencePaths;

        //remove default import specifier
        path.remove();
        if (path.parent.specifiers.length === 0) {
          path.parentPath.remove();
        }

        if (!referencePaths.length) return;

        const configHash = getHash(
          fs.readFileSync(configPath, { encoding: "utf8" })
        );
        if ($configHash !== configHash) {
          $configHash = configHash;
          const {
            resolvedConfig,
            utilitiesRoot,
            componentsRoot,
            baseRoot,
            screens,
            variants,
            generateTwClassSubstituteRoot,
          } = tailwindData(getTailwindConfig(configPath), corePlugins);

          $twClassDictionary = createTwClassDictionary(
            componentsRoot,
            utilitiesRoot
          );

          if (includeBase) {
            const baseCSSroot = baseRoot.clone();
            baseCSSroot.walkAtRules("layer", (atRule) => {
              atRule.replaceWith(atRule.nodes);
            });
            baseCSS = baseCSSroot.toString();
          }

          $compare = (
            [firstClass, firstVariants]: [string, string[]],
            [secondClass, secondVariants]: [string, string[]]
          ) => {
            const firstIsScreen = screens.indexOf(firstVariants[0]);
            const secondIsScreen = screens.indexOf(secondVariants[0]);
            if (firstIsScreen !== -1 || secondIsScreen !== -1) {
              if (firstIsScreen < secondIsScreen) {
                return -1;
              }
              if (firstIsScreen > secondIsScreen) {
                return 1;
              }
              return 0;
            }

            if (firstClass < secondClass) {
              return -1;
            }

            if (firstClass > secondClass) {
              return 1;
            }
            return 0;
          };

          $twClassesUtils = initTwClassesUtils(resolvedConfig.separator, [
            ...screens,
            ...variants,
          ]);
          $generateTwClassSubstituteRoot = generateTwClassSubstituteRoot;
          $generatedTwClassesCSS = {};
        }

        if (!$twClassesUtils) throw new Error("No class utils found");

        const referencedTwClasses = [];
        for (const referencePath of referencePaths) {
          if (referencePath.type !== "Identifier") {
            throw new Error("Reference path is not identifier");
          }

          if (referencePath.parent.type === "TaggedTemplateExpression") {
            const quasi = referencePath.parent.quasi;
            const templateElements = quasi.quasis.map((item) => item.value.raw);
            const expressions = quasi.expressions.map((item) => {
              //@ts-expect-error
              return item.evaluate().value;
            });
            const twTemplateClasses: string[] = [];
            while (templateElements.length || expressions.length) {
              const twClassString = templateElements.shift();
              const twClassObject = expressions.shift();
              if (twClassString) {
                twTemplateClasses.push(twClassString);
              }
              if (twClassObject) {
                twTemplateClasses.push(twClassObject);
              }
            }
            const serializedTwClasses = $twClassesUtils.serializer(
              twTemplateClasses
            );
            referencePath.parentPath.replaceWith(
              t.stringLiteral(serializedTwClasses)
            );
            referencedTwClasses.push(serializedTwClasses);
          }
        }

        if (
          !outputPath ||
          !$generateTwClassSubstituteRoot ||
          !$twClassDictionary
        )
          return;

        let twClasses = $twClassesUtils.composer(referencedTwClasses);
        if (!twClasses.length) return;

        let parsedTwClasses: ParsedTwClass[];
        if (enableCache && state.filename && cacheUtils && cache) {
          //@future https://github.com/babel/babel/pull/11741
          // babel.addExternalDependency(configPath, state.filename)

          const cacheEntry: TailwincssinjsCacheEntry = {
            sourcePath: state.filename,
            sourceHash: getHash(state.file.code),
            twClasses,
          };
          const updatedCache = cacheUtils.updateCache(cache, cacheEntry);
          cacheUtils.writeCache(updatedCache);

          parsedTwClasses = $twClassesUtils.parser(
            updatedCache.map((entry) => entry.twClasses)
          );
        } else {
          parsedTwClasses = $twClassesUtils.parser(twClasses);
        }

        const sortedTwClasses = parsedTwClasses.sort($compare);
        const combinedRoot = [];
        for (const twClass of sortedTwClasses) {
          const generatedTwClass = $twClassesUtils.generator([twClass]);
          let generatedTwClassCSS = $generatedTwClassesCSS[generatedTwClass[0]];
          if (generatedTwClass) {
            generatedTwClassCSS = $generateTwClassSubstituteRoot(
              $twClassDictionary,
              twClass
            ).toString();
          }
          combinedRoot.push(generatedTwClassCSS);
        }
        fs.writeFileSync(
          outputPath,
          [commentCSS, baseCSS, ...combinedRoot].join("\n\n")
        );
      },
    },
  };
}
