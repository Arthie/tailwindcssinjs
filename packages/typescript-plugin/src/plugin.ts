import {
  decorateWithTemplateLanguageService,
  Logger,
  TemplateContext,
  TemplateLanguageService,
} from "typescript-template-language-service-decorator";
import * as ts from "typescript/lib/tsserverlibrary";
import * as vscode from "vscode-languageserver-types";
import { TwObject } from "@tailwindcssinjs/transformers";
import { TailwindConfig } from "@tailwindcssinjs/tailwindcss-data/lib/tailwindcssConfig";

import tailwindcssinjs, { requireTailwindConfig } from "./tailwindcssinjs";
import { createLogger } from "./logger";
import {
  TailwindContext,
  getTailwindContextFromPosition,
  getTailwindContextFromTemplateContext,
} from "./tailwindContext";
import {
  translateCompletionItemsToCompletionEntryDetails,
  translateCompletionEntry,
  translateHoverItemsToQuickInfo,
} from "./translate";
import postcss from "postcss";

let tailwindConfig: TailwindConfig | undefined;
let tailwindData:
  | {
      twObjectMap: Map<string, TwObject>;
      screens: string[];
      variants: string[];
      generateCssFromText: (text: string) => postcss.Root[];
    }
  | undefined;

let watcher: ts.FileWatcher | undefined;

function updateTailwindcssinjs(logger: Logger, configPath?: string) {
  try {
    logger.log(`CONFIG PATH:${configPath}`);
    if (configPath) {
      if (watcher) {
        watcher.close();
      }
      watcher = ts.sys.watchFile?.(
        configPath,
        (fileName: string, eventKind: ts.FileWatcherEventKind) => {
          logger.log(`Watchfile config update: ${configPath}`);
          updateTailwindcssinjs(logger, fileName);
        }
      );
    }
    tailwindConfig = requireTailwindConfig(logger, configPath);
    tailwindData = tailwindcssinjs(tailwindConfig);
  } catch (err) {
    logger.log(`ERROR: ${err}`);
  }
}

function tailwindcssinjsLanguageService(
  info: ts.server.PluginCreateInfo,
  logger: Logger
): TemplateLanguageService {
  try {
    updateTailwindcssinjs(logger, info.config?.config);

    return {
      getSyntacticDiagnostics(context: TemplateContext): ts.Diagnostic[] {
        if (!tailwindData) {
          return [];
        }
        const maxNumberOfProblems = 100;
        let problems = 0;
        const diagnostics: ts.Diagnostic[] = [];

        //TODO improve this regex
        const NESTED_VARIANT_REGEXP = /(\[([^\[\]]){0,}\[[^\]]{0,}\][^\]]{0,}\])|(\[([^\[\]]){0,}\[)|(\]([^\[\]]){0,}\])/;

        const nestedMath = context.text.match(NESTED_VARIANT_REGEXP);

        if (nestedMath?.length) {
          problems++;
          let diagnostic: ts.Diagnostic = {
            file: context.node.getSourceFile(),
            code: 1,
            category: ts.DiagnosticCategory.Error,
            start: nestedMath.index,
            length: nestedMath[0].length,
            messageText: `Nested variants arrays are not allowed.`,
            source: "tailwindcssinjs",
          };
          diagnostics.push(diagnostic);
        }

        const tailwindContext = getTailwindContextFromTemplateContext(
          context,
          tailwindConfig?.separator ?? ":"
        );

        for (const templateContextClass of tailwindContext) {
          if (problems > maxNumberOfProblems) break;

          if (templateContextClass.type === "array") {
            for (const twClass of templateContextClass.classes) {
              const twObject = tailwindData.twObjectMap.get(twClass.text);
              if (!twObject) {
                diagnostics.push({
                  file: context.node.getSourceFile(),
                  code: 2,
                  category: ts.DiagnosticCategory.Error,
                  start: twClass.index,
                  length: twClass.text.length,
                  messageText: `${twClass.text} is not a tailwind class.`,
                  source: "tailwindcssinjs",
                });
                problems++;
              }
            }
          }

          if (templateContextClass.type === "variant") {
            const twObject = tailwindData.twObjectMap.get(
              templateContextClass.class.text
            );
            if (!twObject) {
              diagnostics.push({
                file: context.node.getSourceFile(),
                code: 2,
                category: ts.DiagnosticCategory.Error,
                start: templateContextClass.class.index,
                length: templateContextClass.class.text.length,
                messageText: `${templateContextClass.class.text} is not a tailwind class.`,
                source: "tailwindcssinjs",
              });
              problems++;
            }
          }

          if (templateContextClass.type === "class") {
            if (!tailwindData.twObjectMap.has(templateContextClass.text)) {
              diagnostics.push({
                file: context.node.getSourceFile(),
                code: 5,
                category: ts.DiagnosticCategory.Error,
                start: templateContextClass.index,
                length: templateContextClass.text.length,
                messageText: `${templateContextClass.text} is not a tailwind class.`,
                source: "tailwindcssinjs",
              });
              problems++;
            }
          }

          if (
            templateContextClass.type === "array" ||
            templateContextClass.type === "variant"
          ) {
            //first:second:...other
            const [first, second, ...other] = templateContextClass.variant;

            const getVariantType = (variant?: string) => {
              if (variant) {
                if (tailwindData?.screens.includes(variant)) {
                  return "screen";
                }
                if (tailwindData?.variants.includes(variant)) {
                  return "variants";
                }
              }
            };

            const firstType = getVariantType(first);
            const secondType = getVariantType(second);

            // if (other.length) {
            //   diagnostics.push({
            //     file: context.node.getSourceFile(),
            //     code: 6,
            //     category: ts.DiagnosticCategory.Error,
            //     start: templateContextClass.index,
            //     length: templateContextClass.variant.join(
            //       tailwindConfig?.separator ?? ":"
            //     ).length,
            //     messageText: `${templateContextClass.variant.join(
            //       tailwindConfig?.separator ?? ":"
            //     )} is not a valid tailwind variant. The selector has more than 2 variants: ${other}.`,
            //     source: "tailwindcssinjs",
            //   });
            //   problems++;
            //   continue;
            // }

            // if (secondType && firstType === "variants") {
            //   diagnostics.push({
            //     file: context.node.getSourceFile(),
            //     code: 7,
            //     category: ts.DiagnosticCategory.Error,
            //     start: templateContextClass.index,
            //     length: templateContextClass.variant.join(
            //       tailwindConfig?.separator ?? ":"
            //     ).length,
            //     messageText: `${templateContextClass.variant.join(
            //       tailwindConfig?.separator ?? ":"
            //     )} is not a valid tailwind variant. "${first}" should be a screen variant`,
            //     source: "tailwindcssinjs",
            //   });
            //   problems++;
            //   continue;
            // }

            if (!firstType && !secondType) {
              diagnostics.push({
                file: context.node.getSourceFile(),
                code: 8,
                category: ts.DiagnosticCategory.Error,
                start: templateContextClass.index,
                length: templateContextClass.variant.join(
                  tailwindConfig?.separator ?? ":"
                ).length,
                messageText: `${templateContextClass.variant.join(
                  tailwindConfig?.separator ?? ":"
                )} is not a tailwind variant.`,
                source: "tailwindcssinjs",
              });
              problems++;
              continue;
            }

            if (firstType === "screen" && secondType === "screen") {
              diagnostics.push({
                file: context.node.getSourceFile(),
                code: 9,
                category: ts.DiagnosticCategory.Error,
                start: templateContextClass.index,
                length: templateContextClass.variant.join(
                  tailwindConfig?.separator ?? ":"
                ).length,
                messageText: `${templateContextClass.variant.join(
                  tailwindConfig?.separator ?? ":"
                )} is not a valid tailwind variant. "${second}" should be a variant, not a screen variant`,
                source: "tailwindcssinjs",
              });
              problems++;
              continue;
            }
          }
        }

        return diagnostics;
      },
      getCompletionsAtPosition(
        context: TemplateContext,
        position: ts.LineAndCharacter
      ): ts.CompletionInfo {
        if (!tailwindData) {
          return {
            isGlobalCompletion: false,
            isMemberCompletion: false,
            isNewIdentifierLocation: false,
            entries: [],
          };
        }

        const tailwindContexts = getTailwindContextFromTemplateContext(
          context,
          tailwindConfig?.separator ?? ":"
        );

        const twContext = getTailwindContextFromPosition(
          tailwindContexts,
          position
        );

        //todo add check for is tailwind class
        const entries: ts.CompletionEntry[] = [];
        for (const key of tailwindData.twObjectMap.keys()) {
          const entry = translateCompletionEntry({
            label: key,
            kind: vscode.CompletionItemKind.Text,
          });

          if (twContext) {
            if (twContext.type === "array") {
              const twContextClass = getTailwindContextFromPosition(
                twContext.classes,
                position
              );
              if (twContextClass) {
                entry.replacementSpan = {
                  start: twContextClass.index,
                  length: twContextClass.text.length,
                };
              }
            } else {
              const endsWithSeparator = twContext.text.endsWith(
                tailwindConfig?.separator ?? ":"
              );
              if (!endsWithSeparator) {
                entry.replacementSpan = {
                  start: twContext.index,
                  length: twContext.text.length,
                };
              }
            }
          }
          entries.push(entry);
        }

        for (const variant of [
          ...tailwindData.screens,
          ...tailwindData.variants,
        ]) {
          const variantEntry = translateCompletionEntry({
            label: variant,
            kind: vscode.CompletionItemKind.Function,
          });
          entries.push(variantEntry);
        }

        return {
          isGlobalCompletion: false,
          isMemberCompletion: false,
          isNewIdentifierLocation: false,
          entries: entries,
        };
      },
      getCompletionEntryDetails(
        context: TemplateContext,
        position: ts.LineAndCharacter,
        name: string
      ): ts.CompletionEntryDetails {
        if (!tailwindData) {
          return {
            displayParts: [],
            kind: ts.ScriptElementKind.unknown,
            kindModifiers: "",
            name,
          };
        }

        const twObject = tailwindData.twObjectMap.get(name);
        if (twObject) {
          const root = postcss.root().append(...twObject.nodes);
          return translateCompletionItemsToCompletionEntryDetails({
            label: name,
            kind: vscode.CompletionItemKind.Text,
            documentation: {
              kind: vscode.MarkupKind.Markdown,
              value: [
                "```css",
                root.toString().replace(/    /g, "  "),
                "```",
              ].join("\n"),
            },
          });
        }
        return {
          displayParts: [],
          kind: ts.ScriptElementKind.unknown,
          kindModifiers: "",
          name,
        };
      },
      getQuickInfoAtPosition(
        context: TemplateContext,
        position: ts.LineAndCharacter
      ): ts.QuickInfo | undefined {
        if (!tailwindData) {
          return;
        }

        const tailwindContexts = getTailwindContextFromTemplateContext(
          context,
          tailwindConfig?.separator ?? ":"
        );

        const twContext = getTailwindContextFromPosition(
          tailwindContexts,
          position
        );

        const getQuickInfo = (
          twContext: TailwindContext,
          twObjectRoots: postcss.Root[]
        ) => {
          return translateHoverItemsToQuickInfo({
            start: twContext.index,
            label: twContext.text,
            kind: vscode.CompletionItemKind.Text,
            documentation: {
              kind: vscode.MarkupKind.Markdown,
              value: [
                "```css",
                ...twObjectRoots.map((twObjectRoot: postcss.Root) => [
                  twObjectRoot.toString().replace(/    /g, "  "),
                ]),
                "```",
              ].join("\n"),
            },
          });
        };

        if (twContext) {
          const twObjectRoots = tailwindData.generateCssFromText(
            twContext.text
          );
          return getQuickInfo(twContext, twObjectRoots);
        }
      },
    };
  } catch (err) {
    logger.log(`ERROR: ${err}`);
    throw err;
  }
}

interface PluginConfig {
  config: string;
}

export = (mod: { typescript: typeof ts }): ts.server.PluginModule => {
  let logger: Logger;
  return {
    create(info: ts.server.PluginCreateInfo): ts.LanguageService {
      logger = createLogger(info);
      logger.log("TailwindcssinjsTsPlugin is starting.");
      return decorateWithTemplateLanguageService(
        mod.typescript,
        info.languageService,
        info.project,
        tailwindcssinjsLanguageService(info, logger),
        { tags: ["tw"] },
        { logger }
      );
    },
    onConfigurationChanged(pluginConfig: PluginConfig) {
      if (logger) {
        logger.log(
          `Config has changed ${typeof pluginConfig} ${JSON.stringify(
            pluginConfig
          )}`
        );
      }
      updateTailwindcssinjs(logger, pluginConfig.config);
    },
  };
};
