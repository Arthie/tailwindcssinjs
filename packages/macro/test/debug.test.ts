import pluginTester from "babel-plugin-tester";
import plugin from "babel-plugin-macros";

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {
    filename: __filename,
  },
  tests: {
    "no usage": `import tw from '../lib/macro'`,
    "correct usage 1": `
      import tw from '../lib/macro';
      const css = tw("dark:bg-black bg-gradient-to-r sm:motion-reduce:translate-y-0 motion-reduce:hover:translate-y-0 md:motion-reduce:hover:translate-y-0");
    `,
    "correct usage 5": `
    import tw from '../lib/macro';
    const css = tw("group-hover:bg-red-300");
    `,
    "correct usage 6": `
    import tw from '../lib/macro';
    const css = tw("dark:bg-red-300");
    `,
  },
});
