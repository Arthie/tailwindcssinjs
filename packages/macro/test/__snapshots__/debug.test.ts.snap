// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tailwindcssinjs correct usage 1: correct usage 1 1`] = `

import tw from '../lib/macro';
const css = tw("dark:bg-black bg-gradient-to-r sm:motion-reduce:translate-y-0 motion-reduce:hover:translate-y-0 md:motion-reduce:hover:translate-y-0");

      ↓ ↓ ↓ ↓ ↓ ↓

const css = {
  backgroundImage: "linear-gradient(to right, var(--gradient-color-stops))",
  "@media (min-width: 640px)": {
    "@media (prefers-reduced-motion: reduce)": {
      "--transform-translate-y": "0",
    },
  },
  "@media (min-width: 1024px)": {
    "@media (prefers-reduced-motion: reduce)": {
      "&:hover": {
        "--transform-translate-y": "0",
      },
    },
  },
  "@media (prefers-color-scheme: dark)": {
    "--bg-opacity": "1",
    backgroundColor: ["#000000", "rgba(0, 0, 0, var(--bg-opacity))"],
  },
  "@media (prefers-reduced-motion: reduce)": {
    "&:hover": {
      "--transform-translate-y": "0",
    },
  },
};


`;

exports[`tailwindcssinjs correct usage 5: correct usage 5 1`] = `

import tw from '../lib/macro';
const css = tw("group-hover:bg-red-300");

      ↓ ↓ ↓ ↓ ↓ ↓

const css = {
  ".group:hover &": {
    "--bg-opacity": "1",
    backgroundColor: ["#f8b4b4", "rgba(248, 180, 180, var(--bg-opacity))"],
  },
};


`;

exports[`tailwindcssinjs correct usage 6: correct usage 6 1`] = `

import tw from '../lib/macro';
const css = tw("dark:bg-red-300");

      ↓ ↓ ↓ ↓ ↓ ↓

const css = {
  "@media (prefers-color-scheme: dark)": {
    "--bg-opacity": "1",
    backgroundColor: ["#f8b4b4", "rgba(248, 180, 180, var(--bg-opacity))"],
  },
};


`;

exports[`tailwindcssinjs no usage: no usage 1`] = `

import tw from '../lib/macro'

      ↓ ↓ ↓ ↓ ↓ ↓



`;
